WIP: save learning rate scheduler to training state

epwalsh opened the pull request:
When using a learning rate scheduler, I think user probably expects the scheduler to pick up where it left off when restoring a trainer from a checkpoint. This makes sure that the scheduler state is stored with each checkpoint when a scheduler is used. Let me know if you have any comments!

**TODO:**
- [x] modify `Trainer._restore_checkpoint`
- [x] modify `Trainer._save_checkpoint`
- [ ] add unit tests for `Trainer`
- [ ] add unit tests for `CosineWithRestarts` scheduler to make sure state is correctly restored

matt-gardner commented:
@DeNeutoy, you mind taking this one?  I've never touched the checkpointing code; I think you have more context there.

epwalsh commented:
Just a heads up, I might have just found a bug w/ the restoring from checkpoint for learning rate schedulers. I'm looking into it now. Changing the title back to "WIP"

epwalsh commented:
Update: the bug was because the checkpoints were being saved before updating the learning rate scheduler, so when the learning rate scheduler was restored, it was an epoch behind. I fixed this and added a test for it in `tests.training.trainer_test.py` 😅
